Cannot filter by "testnames" when suite xml is a suite of suites

krmahadevan opened the pull request:
Closes #1594

As part of this changeset, extracted out the 
unrelated xmlsuite manipulation code out of TestNG
class and then added them to XmlSuite

Fixes #1594 .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

krmahadevan commented on src/main/java/org/testng/xml/XmlSuite.java:
```diff
@@ -1017,5 +1019,114 @@ public void addTest(XmlTest test) {
     return result;
   }
+
+  private XmlSuite cleanClone(List<XmlTest> tests) {
```
I was just piggy backing on the fact that it already holds logic such as `shallowCopy()` and `clone()`

juherr replied:
XmlObjects should just be POJO for the xml binding without business logic, could you move these methods in an external util class?

krmahadevan commented on src/main/java/org/testng/xml/XmlSuite.java:
```diff
@@ -1017,5 +1019,114 @@ public void addTest(XmlTest test) {
+
+  /**
+   * Creates a cloned copy of the current {@link XmlSuite} if the current suite
+   * contains atleast on &lt;test&gt; whose name matches with the provided names.
```
Fixed.

juherr replied:
at least?

krmahadevan commented on src/main/java/org/testng/xml/XmlSuite.java:
```diff
@@ -1017,5 +1019,114 @@ public void addTest(XmlTest test) {
+      String name = suite.getName();
+
+      int count = 0;
+      String tmpName = name;
```
Why do we need to use `StringBuilder` here ? We are not building the string in the loop. We are just overwriting its value with the previous value. A simple concatenation of strings is literally equal to what the StringBuilder does especially when the concatenation is not being done to construct a string in a loop.

juherr replied:
StringBuilder possible here?

krmahadevan commented on src/main/java/org/testng/xml/XmlSuite.java:
```diff
@@ -1017,5 +1019,114 @@ public void addTest(XmlTest test) {
+        suite.setName(tmpName);
+        names.add(tmpName);
+      } else {
+        names.add(name);
```
This is ported logic. Not newly written.

juherr replied:
You add name twice in that case. Is it expected?

krmahadevan commented on src/main/java/org/testng/xml/XmlTest.java:
```diff
@@ -752,4 +752,17 @@ public void setGroups(XmlGroups xmlGroups) {
+   * @param names The list of names to check.
+   * @return <code>true</code> if the current test's name matches with any of the given names.
+   */
+  public boolean nameMatchesAny(List<String> names) {
```
I think for this case, we can skip having to create a separate utility class because this is effectively speaking not a business logic, but just something that works with an internal attribute. So I am going to leave this as is.

juherr replied:
XmlObjects should just be POJO for the xml binding without business logic, could you move this method in an external util class?

krmahadevan commented on src/main/java/org/testng/xml/XmlTest.java:
```diff
@@ -752,4 +752,17 @@ public void setGroups(XmlGroups xmlGroups) {
+   * @return <code>true</code> if the current test's name matches with any of the given names.
+   */
+  public boolean nameMatchesAny(List<String> names) {
+    for (String name : names) {
```
Fixed.

juherr replied:
Maybe just
```java
return names.contains(getName());
```

krmahadevan commented on src/test/java/org/testng/xml/XmlSuiteTest.java:
```diff
@@ -11,15 +11,17 @@
         suite.addExcludedGroup("bar");
-        assertEquals(Collections.singletonList("foo"), suite.getIncludedGroups());
-        assertEquals(Collections.singletonList("bar"), suite.getExcludedGroups());
+        assertThat(suite.getIncludedGroups()).hasSameElementsAs(Collections.singletonList("foo"));
```
Fixed.

juherr replied:
You can avoid the singleton list because assertj provides methods the check on elements

krmahadevan commented on src/test/java/org/testng/xml/XmlSuiteTest.java:
```diff
@@ -42,13 +44,74 @@ public void testTimeOut(String timeout, int size, int lineNumber) throws IOExcep
         List<Integer> lineNumbers = grep(stringReader, "time-out=\"1000\"", resultLines);
-        assertEquals(lineNumbers.size(), size);
-        assertEquals(resultLines.size(), size);
+        assertThat(lineNumbers.size()).isEqualTo(size);
```
fixed.

juherr replied:
assertj provides an assertion on list size

krmahadevan commented on src/test/java/org/testng/xml/XmlSuiteTest.java:
```diff
@@ -42,13 +44,74 @@ public void testTimeOut(String timeout, int size, int lineNumber) throws IOExcep
+    }
+
+    @Test(expectedExceptions = TestNGException.class,
+    expectedExceptionsMessageRegExp = "\nThe test\\(s\\) \\<\\[test3\\]\\> cannot be found.")
```
Have applied formatting from the IDE and still this is how it is. So i guess the indentation is proper.

juherr replied:
Indentation

krmahadevan commented on src/test/java/test/SimpleBaseTest.java:
```diff
@@ -120,8 +120,16 @@ protected static TestNG createTests(Path outDir,String suiteName, Class<?>... te
+  protected static XmlSuite createDummySuiteWithTestNamesAs(String ... tests) {
+    XmlSuite suite = new XmlSuite();
+    suite.setName("random_suite");
+    for (String test: tests) {
```
Fixed.

juherr replied:
space before ':'

juherr commented on src/main/java/org/testng/xml/XmlSuite.java:
```diff
@@ -1017,5 +1019,114 @@ public void addTest(XmlTest test) {
+      String name = suite.getName();
+
+      int count = 0;
+      String tmpName = name;
```
true, forget it. (strange, github shows comments in the wrong order)

juherr commented on src/main/java/org/testng/xml/XmlSuite.java:
```diff
@@ -1017,5 +1019,114 @@ public void addTest(XmlTest test) {
+        suite.setName(tmpName);
+        names.add(tmpName);
+      } else {
+        names.add(name);
```
ok, but you can remove the else case because it is useless IMO

juherr commented on src/main/java/org/testng/xml/XmlTest.java:
```diff
@@ -752,4 +752,17 @@ public void setGroups(XmlGroups xmlGroups) {
+   * @param names The list of names to check.
+   * @return <code>true</code> if the current test's name matches with any of the given names.
+   */
+  public boolean nameMatchesAny(List<String> names) {
```
debatable but ok

juherr commented on src/test/java/org/testng/xml/XmlSuiteTest.java:
```diff
@@ -42,13 +44,74 @@ public void testTimeOut(String timeout, int size, int lineNumber) throws IOExcep
+    }
+
+    @Test(expectedExceptions = TestNGException.class,
+    expectedExceptionsMessageRegExp = "\nThe test\\(s\\) \\<\\[test3\\]\\> cannot be found.")
```
strange because the previous method has another indentation

krmahadevan commented:
@juherr - I have fixed all your observations. To facilitate ease of reviews, I amended the commit and pushed in the changes again. Please help take a look. I have now extracted out all the new logic into `XmlSuiteUtils` into the internal package. Please check.

juherr commented on src/test/java/org/testng/xml/XmlSuiteTest.java:
```diff
@@ -42,10 +42,10 @@ public void testTimeOut(String timeout, int size, int lineNumber) throws IOExcep
         List<Integer> lineNumbers = grep(stringReader, "time-out=\"1000\"", resultLines);
-        assertEquals(lineNumbers.size(), size);
-        assertEquals(resultLines.size(), size);
+        assertThat(lineNumbers).size().isEqualTo(size);
```
`assertThat(lineNumbers).hasSize(size);`

krmahadevan commented:
ping @cbeust - Can you please merge if you are fine with the changes ?
